bindings/pyroot/pythonizations/python/ROOT/_pythonization/_tmva/_gnn.py

```diff
@@ -0,0 +1,259 @@
+                else:
+                    print("Invalid Model for node update.")
+                    return
+        
+        else:
+            print("Invalid Model for node update.")
+            return
+        
+        add_weights(node_model.variables, gin.nodes_update_block)
+
+        # adding the edge update function
+        edge_model = GraphModule._edge_block._edge_model
+        if (edge_model.name == 'mlp'):
+            make_mlp_model(gin, edge_model, gbl_namespace.TMVA.Experimental.SOFIE.FunctionTarget.EDGES, gbl_namespace.TMVA.Experimental.SOFIE.GraphType.GNN)
+        elif (edge_model.name == 'sequential'):
+            for i in edge_model._layers:
+                if(i.name == 'mlp'):
+                    make_mlp_model(gin, edge_model._layers[0], gbl_namespace.TMVA.Experimental.SOFIE.FunctionTarget.EDGES, gbl_namespace.TMVA.Experimental.SOFIE.GraphType.GNN)
+                elif(i.name == 'layer_norm'):
+                    add_layer_norm(i, gin.edges_update_block)
+                else:
+                    print("Invalid Model for edge update.")
+                    return
+        
+        else:
+            print("Invalid Model for edge update.")
+            return
+
+        add_weights(edge_model.variables, gin.edges_update_block)
+
+        # adding the global update function
+        global_model = GraphModule._global_block._global_model
+        if (global_model.name == 'mlp'):
+            make_mlp_model(gin, global_model, gbl_namespace.TMVA.Experimental.SOFIE.FunctionTarget.GLOBALS, gbl_namespace.TMVA.Experimental.SOFIE.GraphType.GNN)
+        elif (global_model.name == 'sequential'):
+            for i in global_model._layers:
+                if(i.name == 'mlp'):
+                    make_mlp_model(gin, global_model._layers[0], gbl_namespace.TMVA.Experimental.SOFIE.FunctionTarget.GLOBALS, gbl_namespace.TMVA.Experimental.SOFIE.GraphType.GNN)
+                elif(i.name == 'layer_norm'):
+                    add_layer_norm(i, gin.globals_update_block)
+                else:
+                    print("Invalid Model for global update.")
+                    return
+        
+        else:
+            print("Invalid Model for global update.")
+            return
+
+        add_weights(global_model.variables, gin.globals_update_block)
+
+        # adding edge-node aggregate function
+        add_aggregate_function(gin, GraphModule._node_block._received_edges_aggregator._reducer.__qualname__, gbl_namespace.TMVA.Experimental.SOFIE.FunctionRelation.NODES_EDGES)
+
+        # adding node-global aggregate function
+        add_aggregate_function(gin, GraphModule._global_block._nodes_aggregator._reducer.__qualname__, gbl_namespace.TMVA.Experimental.SOFIE.FunctionRelation.NODES_GLOBALS)
+
+        # adding edge-global aggregate function
+        add_aggregate_function(gin, GraphModule._global_block._edges_aggregator._reducer.__qualname__, gbl_namespace.TMVA.Experimental.SOFIE.FunctionRelation.EDGES_GLOBALS)
```

Comment:
Here you are accessing a lot of private attributes, e.g. `_global_block`, `_edges_aggregator` etc. This is usually not a good pattern. Are you responsible for the class type of `GraphModule` argument? Are you sure that those private attributes will not change? If the class is defined in the C++ part of this PR, you should think about making those attributes private data members and then creating public getters

Reply:
The private attributes are actually from the Graph modules defined in `graph_nets`. They are required while parsing to extract information about the passed GNN or GraphIndependent model to instantiate the corresponding `RModel` objects. We don't have the required public getters (that I am aware of) defined in those `graph_nets` models that we can use to access them.